### PR TITLE
pypeit_collate_1d improvements and refactor

### PR DESCRIPTION
**Pull request by badpandabear**
These changes include:

Refactoring the collating code into a collate.py module under pypeit.core

Refactoring the archive to separate out KOA specific portions. 

Refactoring some of the config file parsing code from pypeit_collate_1d to par.util

Renamed exclude_slits to make it more clear it excludes based on bitmask.

Added --exclude_serendip based on userfeedback

Improve error handling so that pypeit_collate_1d doesn't exit when a SpecObj does not have the right OPT_COUNTS/BOX_COUNTS for coadding.

### CODE REVIEW

**Comment by badpandabear** [comment]
Unit tests:

> ====================================== 277 passed, 423 warnings in 798.21s (0:13:18) ============================

Dev Suite:

> --- PYPEIT DEVELOPMENT SUITE PASSED 110/110 TESTS (Masters ignored) ---
> Testing Started at 2021-05-12T11:19:07.436103
> Testing Completed at 2021-05-12T16:11:36.981523
> Total Time: 4:52:29.545420
>

File: pypeit/archive.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
+"""
```
Comment (profxj) [inline pypeit/archive.py:1]: Thanks for doing this!

File: pypeit/core/collate.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
+#
```
Comment (profxj) [inline pypeit/core/collate.py:1]: and this too :)

File: pypeit/par/util.py [hunk 3]
```diff
@@ -432,6 +432,56 @@ def _parse_setup_lines(lines):
     #
     return setups
 
+def parse_tool_config(config_file, block, check_files=False):
```
Comment (profxj) [inline pypeit/par/util.py:435]: nice!  I need to implement this for DEIMOS 2D coadd

File: pypeit/tests/files/ipac/by_id_meta.dat [hunk 4]
```diff
@@ -1,6 +1,5 @@
...
-|                  char|                                                               char|           double|            double|       char|   char|    char|  char|    char|   char|   char|      double|    double| double|
-|                      |                                                                   |                 |                  |           |       |        |      |        |       |       |            |          |       |
-|                  null|                                                               null|             null|              null|       null|   null|    null|  null|    null|   null|   null|        null|      null|   null|
- DE.20100913.22358.fits    spec1d_DE.20100913.22358-CFHQS1_DEIMOS_2010Sep13T061231.334.fits           352.304 -3.097083333333333      CFHQS1    None     None   None     830G CFHQS1B     1,1 55452.258696 2.40403914  1800.0 
- KB.20191219.56886.fits spec2d_KB.20191219.56886-BB1245p4238_KCWI_2019Dec19T154806.538.fits 191.3984583333333  42.64016666666667 BB1245p4238    None    2019B   U190      BH2   Small     1,1 58836.658409       1.15  1200.0 
+|  id|                                                 file|      date|
```
Comment (profxj) [inline pypeit/tests/files/ipac/by_id_meta.dat]: warning -- I touched this in `odds_and_ends`
a conflict is likely coming..

File: pypeit/tests/test_collate_1d.py [hunk 5]
```diff
@@ -1,5 +1,5 @@
 """
```
Comment (profxj) [inline pypeit/tests/test_collate_1d.py:1]: I touched this too, but I think the merge will go ok.

**Review by profxj — APPROVED** [review]
Good stuff.   Mainly minor comments for me
so I am approving now.

**Review by debora-pe — APPROVED** [review]
Looks good to me!

**Pull request merged**